firedrake/slate/slac/compiler.py

```diff
@@ -41,18 +45,167 @@
+                    tensor = eigen_tensor(exp, self.temps[exp], indices)
+                    call = ast.FunCall(kinfo.kernel.name,
+                                       tensor,
+                                       coord_sym,
+                                       *args)
+                    assembly_calls[it_type].append(call)
+
+                    # Subkernels for local assembly (Eigen templated functions)
+                    kast = transformer.visit(kinfo.kernel._ast)
+                    templated_subkernels.append(kast)
+                    include_dirs.extend(kinfo.kernel._include_dirs)
+                    oriented = oriented or kinfo.oriented
+
+        self.assembly_calls = assembly_calls
+        self.templated_subkernels = templated_subkernels
+        self.include_dirs = list(set(include_dirs))
+        self.oriented = oriented
+
+    @property
+    def integral_type(self):
+        """Returns the integral type associated with a Slate kernel.
+
+        Note that Slate kernels (for now) are always of type 'cell' since
+        these are localized kernels for element-wise linear algebra. This
+        may change in the future if we want Slate to be used for LDG/CDG
+        finite element discretizations.
+        """
+        return "cell"
+
+    @cached_property
+    def needs_cell_facets(self):
+        """Searches for any embedded forms (by inspecting the ContextKernels)
+        which require looping over cell facets. If any are found, this function
+        returns `True` and `False` otherwise.
+        """
+        cell_facet_types = ["interior_facet",
+                            "exterior_facet",
+                            "interior_facet_vert",
+                            "exterior_facet_vert"]
+        return any(cxt_k.original_integral_type in cell_facet_types
+                   for cxt_k in self.context_kernels)
+
+    @cached_property
+    def needs_mesh_layers(self):
+        """Searches for any embedded forms (by inspecting the ContextKernels)
+        which require mesh level information (extrusion measures). If any are
+        found, this function returns `True` and `False` otherwise.
+        """
+        mesh_layer_types = ["interior_facet_horiz_top",
+                            "interior_facet_horiz_bottom",
+                            "exterior_facet_bottom",
+                            "exterior_facet_top"]
+        return any(cxt_k.original_integral_type in mesh_layer_types
+                   for cxt_k in self.context_kernels)
+
+    def construct_ast(self, *args):
+        """Constructs the final kernel AST.
+
+        :arg args: An iterable of macro kernel functions, which call subkernels
+                   and perform elemental linear algebra.
+
+        Returns: The complete kernel AST as a COFFEE `ast.Node`
+        """
+        macro_kernels = list(args)
+        kernel_ast = self.templated_subkernels
+        kernel_ast.extend(macro_kernels)
```

Comment:
This modifies `self.templated_subkernels` in place, is that intentional?

Also confusing is that there's somehow no guarantee that the provided macro kernels only call subkernels that this kernel builder object already knows about.